zerver/tests/test_example.py

```diff
@@ -0,0 +1,318 @@
+        # events.
+        with self.tornado_redirected_to_list(events, expected_num_events=1):
+            result = self.api_post(cordelia, "/api/v1/users/me/status", params)
+
+        self.assert_json_success(result)
+
+        # Check that the POST to Zulip causes the correct events to be sent
+        # to Tornado.
+        self.assertEqual(
+            events[0]["event"],
+            dict(type="user_status", user_id=cordelia.id, status_text="on vacation"),
+        )
+
+        row = UserStatus.objects.last()
+        self.assertEqual(row.user_profile_id, cordelia.id)
+        self.assertEqual(row.status_text, "on vacation")
+
+
+class TestStreamHelpers(ZulipTestCase):
+    # Streams are an important concept in Zulip, and ZulipTestCase
+    # has helpers such as subscribe, users_subscribed_to_stream,
+    # and make_stream.
+    def test_new_streams(self) -> None:
+        cordelia = self.example_user("cordelia")
+        othello = self.example_user("othello")
+        realm = cordelia.realm
+
+        stream_name = "Some new stream"
+        self.subscribe(cordelia, stream_name)
+
+        self.assertEqual(set(self.users_subscribed_to_stream(stream_name, realm)), {cordelia})
+
+        self.subscribe(othello, stream_name)
+        self.assertEqual(
+            set(self.users_subscribed_to_stream(stream_name, realm)), {cordelia, othello}
+        )
+
+    def test_private_stream(self) -> None:
+        # When we test stream permissions, it's very common to use at least
+        # two users, so that you can see how different users are impacted.
+        # We commonly use Othello to represent the "other" user from the primary user.
+        cordelia = self.example_user("cordelia")
+        othello = self.example_user("othello")
+
+        realm = cordelia.realm
+        stream_name = "Some private stream"
+
+        # Use the invite_only flag in make_stream to make a stream "private".
+        stream = self.make_stream(stream_name=stream_name, invite_only=True)
+        self.subscribe(cordelia, stream_name)
+
+        self.assertEqual(set(self.users_subscribed_to_stream(stream_name, realm)), {cordelia})
+
+        stream = get_stream(stream_name, realm)
+        self.assertEqual(stream.name, stream_name)
+        self.assertTrue(stream.invite_only)
+
+        # We will now observe that Cordelia can access the stream...
+        access_stream_for_send_message(cordelia, stream, forwarder_user_profile=None)
+
+        # ...but Othello can't.
+        msg = "Not authorized to send to stream"
+        with self.assertRaisesRegex(JsonableError, msg):
+            access_stream_for_send_message(othello, stream, forwarder_user_profile=None)
+
+
+class TestMessageHelpers(ZulipTestCase):
+    # If you are testing behavior related to messages, then it's good
+    # to know about send_stream_message, send_personal_message, and
+    # most_recent_message.
+    def test_stream_message(self) -> None:
```

Comment:
Should we also advertise/talk about `ZulipTestCase.get_last_message()`? There's a good number of tests which utilize that.

Also, how about consolidating functions like these in either one of `ZulipTestCase` or `test_helpers`?

Reply:
I'd be fine to add `get_last_message` in the examples, since we do use it 97 times.  OTOH I think `most_recent_message` actually is what you commonly want, since `most_recent_message` gets the last message specifically sent to a user.

I am hoping that once we get this merged, folks can iterate on it a bit.

For these two functions in particular, I think they belong in `test_helpers`, since they're not really coupled to the testing framework at all.  I think `ZulipTestCase` is good for functions like `client_get` and `client_post` that are wrappers for Django things and actually require `self` to be able to work.

Reply:
I've generally been preferring to use ZulipTestCase as home for useful testing utility functions, in part because having two places for things means it's harder for folks to find things, and `test_helpers.py` is definitely the secondary place for the reasons noted above.